src/plugins/discover/public/application/doc_views/doc_views_types.ts

```diff
@@ -49,17 +49,34 @@ export type DocViewRenderFn = (
   renderProps: DocViewRenderProps
 ) => () => void;
 
-export interface DocViewInput {
-  component?: DocViewerComponent;
-  directive?: AngularDirective;
+export interface BaseDocViewInput {
   order: number;
-  render?: DocViewRenderFn;
   shouldShow?: (hit: ElasticSearchHit) => boolean;
   title: string;
 }
 
-export interface DocView extends DocViewInput {
-  shouldShow: (hit: ElasticSearchHit) => boolean;
+export interface RenderDocViewInput extends BaseDocViewInput {
+  render: DocViewRenderFn;
+  component?: undefined;
+  directive?: undefined;
 }
 
+interface ComponentDocViewInput extends BaseDocViewInput {
+  component: DocViewerComponent;
+  render?: undefined;
+  directive?: undefined;
+}
+
+interface DirectiveDocViewInput extends BaseDocViewInput {
+  component?: undefined;
+  render?: undefined;
+  directive: ng.IDirective;
+}
+
+export type DocViewInput = ComponentDocViewInput | RenderDocViewInput | DirectiveDocViewInput;
+
+export type DocView = DocViewInput & {
+  shouldShow: NonNullable<DocViewInput['shouldShow']>;
```

Comment:
:information_source: Since I assume we always want to have the same type here, I removed the duplication.